### PR TITLE
fix:(next/image) handle `remotePatterns` with a dot in the pathname

### LINKED ISSUE #46903: Bug in pathname matching in remotePatterns
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: linux
      Arch: x64
      Version: #1 SMP Fri Jan 27 02:56:13 UTC 2023
    Binaries:
      Node: 18.14.1
      npm: 9.3.1
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.2.3
      eslint-config-next: 13.2.3
      react: 18.2.0
      react-dom: 18.2.0
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Image optimization (next/image, next/legacy/image)

### Link to the code that reproduces this issue

https://stackblitz.com/edit/vercel-next-js-vxvpyx?file=pages/index.tsx

### To Reproduce

A url like: https://s.yimg.com/ny/api/res/1.2/.tN3tiOdZqngt0Dro4127w--/YXBwaWQ9aGlnaGxhbmRlcjt3PTEyMDA7aD02NzU-/https://media.zenfs.com/en/total_film_411/be25e8f1a01fff35142f9c4e3d8d215f

fails to match in the otherwise valid remotePatterns

    remotePatterns: [
      {
        protocol: "https",
        hostname: "*.yimg.com",
      },
    ],

### Describe the Bug

When using remotePatterns without a pathname next will default to the following code in match-remote-pattern.ts

```javascript
if (!makeRe(pattern.pathname ?? '**').test(url.pathname)) {
    return false
}
```

This however will fail to match a path in a url like:
https://s.yimg.com/ny/api/res/1.2/.tN3tiOdZqngt0Dro4127w--/YXBwaWQ9aGlnaGxhbmRlcjt3PTEyMDA7aD02NzU-/https://media.zenfs.com/en/total_film_411/be25e8f1a01fff35142f9c4e3d8d215f

Foe now i solved it myself by changing the code to use a different regex
```javascript
let regex = /^(\/[^\s/?#]*)*$/
```

### Expected Behavior

The remote pattern should match

### Which browser are you using? (if relevant)

_No response_

### How are you deploying your application? (if relevant)

_No response_

### LINKED ISSUE #58139: <Image> component crashes when image url lacks filename, e.g. `/.png`
### Link to the code that reproduces this issue

https://github.com/jakst/next-image-error-repro

### To Reproduce

1. Start the dev server
2. Open the app
3. The render crashes with the following error
```
Error: Invalid src prop (https://github.com/.png) on `next/image`, hostname "github.com" is not configured under images in your `next.config.js`
See more info: https://nextjs.org/docs/messages/next-image-unconfigured-host
```

### Current vs. Expected behavior

I expected the page to render, but with a missing image.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: x64
  Version: Darwin Kernel Version 23.0.0: Fri Sep 15 14:42:42 PDT 2023; root:xnu-10002.1.13~1/RELEASE_X86_64
Binaries:
  Node: 21.0.0
  npm: N/A
  Yarn: 4.0.0
  pnpm: 8.10.2
Relevant Packages:
  next: 14.0.2-canary.18
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: N/A
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Image optimization (next/image, next/legacy/image)

### Additional context

This happened when switching from the deprecated images.domains in next.config
```javascript
const nextConfig = {
  images: {
    domains: ["github.com"],
  },
}
```

To the new remotePatterns
```javascript
const nextConfig = {
  images: {
    remotePatterns: [
      {
        protocol: "https",
        hostname: "github.com",
      },
    ],
  },
}
```

They are both valid and contain the "github.com" hostname, so the error message is misleading. But also, the error shouldn't be shown at all because the URL is valid.

### LINKED ISSUE #60483: With all remotePatterns allowed for images, some image URLs are still rejected
### Link to the code that reproduces this issue

https://github.com/WillCalderwood/next-bug

### To Reproduce

1. Start the application and you'll receive the error

```
Error: Invalid src prop [(https://bioage.typepad.com/.a/6a00d8341c4fbe53ef02c8d3a82122200d-600wi)](http://localhost:3000/admin/news/(https://bioage.typepad.com/.a/6a00d8341c4fbe53ef02c8d3a82122200d-600wi)) on `next/image`, hostname "bioage.typepad.com" is not configured under images in your `next.config.js`
See more info: https://nextjs.org/docs/messages/next-image-unconfigured-host
```

### Current vs. Expected behavior

The image `https://bioage.typepad.com/.a/6a00d8341c4fbe53ef02c8d3a82122200d-600wi` is not allowed when all image remotes are set as valid with:
```
 images: {
    remotePatterns: [
      {
        protocol: "https",
        hostname: "**",
      },
      {
        protocol: "http",
        hostname: "**",
      },
    ],
  },
```

The image should not be throwing an error.

Note: It works correctly with `unoptimized={true}` specified.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 10 Pro
Binaries:
  Node: 20.8.1
  npm: N/A
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 14.0.5-canary.47
  eslint-config-next: 14.0.4
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Image optimization (next/image, next/legacy/image)

### Which stage(s) are affected? (Select all that apply)

Vercel (Deployed)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by alexdln**
### What?
Fix remotePatterns when all paths and/or domains are allowed.

### Why?

micromatch creates a very strange regex for all paths - `/^(?:(?!\.)(?:(?:(?!(?:^|[\\/])\.).)*?)[\\/]?)$/`. That is, paths cannot start with a dot or contain a slash followed by a dot.

Interestingly, here are some valid paths:

- /a/a.a/6a00d8341c4fbe53ef02c8d3a82122200d-600wi
- ////a/a.a/6a00d8341c4fbe53ef02c8d3a82122200d-600wi
- ///:?%;№%/a/a.a/6a00d8341c4fbe53ef02c8d3a82122200d-600wi.\/
- /:./6a00d8341c4fbe53ef02c8d3a82122200d-600wi.\/

And here are some invalid ones:

- /.a/6a00d8341c4fbe53ef02c8d3a82122200d-600wi
- /a/.a/6a00d8341c4fbe53ef02c8d3a82122200d-600wi
- ./a/6a00d8341c4fbe53ef02c8d3a82122200d-600wi

I don't think this check makes any sense.

### How?

If the user allows all (`**`) - it means any path or domain will be considered valid.

- Fixes #60483
- Fixes #58139
- Fixes #46903

### CODE REVIEW

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 246a09a8f0c6878a19c82f442ad6d282392c32bc

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 246a09a8f0c6878a19c82f442ad6d282392c32bc

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 246a09a8f0c6878a19c82f442ad6d282392c32bc

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Comment by vordgi** [comment]
I see a test for remotePatterns - test\integration\image-optimizer\test\index.test.ts, but it only checks for schema validation errors, not patterns. In this case, should tests for patterns be added and should they be written in this file?

**Comment by vordgi** [comment]
Maybe remove micromatch and write custom regexp?

Something like:
- ```new RegExp(`^${pathnameRule.replace(/([^a-zA-Z0-9*])/g, '\\$1').replace(/\*\*/g, '.+').replace(/\*/g, '[^\\]+')}$`)``` for pathnames
- ```new RegExp(`^${domainRule.replace(/([^a-zA-Z0-9*])/g, '\\$1').replace(/\*\*/g, '.+').replace(/\*/g, '[^.]+')}$`)``` for domains

`domainRule.replace(/([^a-zA-Z0-9*])/g, '\\$1')` - part for the normalization regular expression, so that the point is checked as a point (`\.`), and not as any character (`.`)

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 246a09a8f0c6878a19c82f442ad6d282392c32bc

`pnpm test-dev test/e2e/app-dir/not-found/basic/index.test.ts`
- app dir - not-found - basic > with runtime = edge > should render the 404 page when the file is removed, and restore the page when re-added

<details>
<summary>Expand output</summary>

  ● app dir - not-found - basic › with runtime = edge › should render the 404 page when the file is removed, and restore the page when re-added

    TIMED OUT: Root Not Found

    My page

    undefined

      626 |
      627 |   if (hardError) {
    > 628 |     throw new Error('TIMED OUT: ' + regex + '\n\n' + content + '\n\n' + lastErr)
          |           ^
      629 |   }
      630 |   return false
      631 | }

      at check (lib/next-test-utils.ts:628:11)
      at Object.<anonymous> (e2e/app-dir/not-found/basic/index.test.ts:106:11)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vordgi/next.js fix-all-remote-patterns | Change |
| - | - | - | - |
| buildDuration | 11.7s | 12.8s | ⚠️ +1.1s |
| buildDurationCached | 6.2s | 6s | N/A |
| nodeModulesSize | 200 MB | 200 MB | N/A |
| nextStartRea..uration (ms) | 425ms | 427ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vordgi/next.js fix-all-remote-patterns | Change |
| - | - | - | - |
| 3f784ff6-HASH.js gzip | 53.4 kB | 53.3 kB | N/A |
| 423.HASH.js gzip | 185 B | 182 B | N/A |
| 68-HASH.js gzip | 29.9 kB | 28.6 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 238 B | 242 B | N/A |
| main-HASH.js gzip | 31.8 kB | 31.8 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | N/A |
| Overall change | 45.2 kB | 45.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vordgi/next.js fix-all-remote-patterns | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vordgi/next.js fix-all-remote-patterns | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 195 B | N/A |
| _error-HASH.js gzip | 182 B | 181 B | N/A |
| amp-HASH.js gzip | 502 B | 502 B | ✓ |
| css-HASH.js gzip | 320 B | 321 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | ✓ |
| edge-ssr-HASH.js gzip | 255 B | 253 B | N/A |
| head-HASH.js gzip | 350 B | 349 B | N/A |
| hooks-HASH.js gzip | 368 B | 369 B | N/A |
| image-HASH.js gzip | 4.18 kB | 4.28 kB | N/A |
| index-HASH.js gzip | 257 B | 256 B | N/A |
| link-HASH.js gzip | 2.61 kB | 2.61 kB | N/A |
| routerDirect..HASH.js gzip | 310 B | 311 B | N/A |
| script-HASH.js gzip | 384 B | 383 B | N/A |
| withRouter-HASH.js gzip | 306 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.11 kB | 3.11 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vordgi/next.js fix-all-remote-patterns | Change |
| - | - | - | - |
| _buildManifest.js gzip | 484 B | 484 B | ✓ |
| Overall change | 484 B | 484 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vordgi/next.js fix-all-remote-patterns | Change |
| - | - | - | - |
| index.html gzip | 527 B | 528 B | N/A |
| link.html gzip | 541 B | 540 B | N/A |
| withRouter.html gzip | 523 B | 523 B | ✓ |
| Overall change | 523 B | 523 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vordgi/next.js fix-all-remote-patterns | Change |
| - | - | - | - |
| edge-ssr.js gzip | 94 kB | 93.8 kB | N/A |
| page.js gzip | 150 kB | 148 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vordgi/next.js fix-all-remote-patterns | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 624 B | 624 B | ✓ |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 37.6 kB | 37.5 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 2.7 kB | 2.7 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vordgi/next.js fix-all-remote-patterns | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 170 kB | 169 kB | N/A |
| app-page-exp..prod.js gzip | 95.8 kB | 95.3 kB | N/A |
| app-page-tur..prod.js gzip | 96.4 kB | 96 kB | N/A |
| app-page-tur..prod.js gzip | 91 kB | 90.6 kB | N/A |
| app-page.run...dev.js gzip | 142 kB | 142 kB | N/A |
| app-page.run..prod.js gzip | 90.3 kB | 89.9 kB | N/A |
| app-route-ex...dev.js gzip | 22.2 kB | 24.1 kB | ⚠️ +1.83 kB |
| app-route-ex..prod.js gzip | 14.9 kB | 16.7 kB | ⚠️ +1.8 kB |
| app-route-tu..prod.js gzip | 14.9 kB | 16.7 kB | ⚠️ +1.8 kB |
| app-route-tu..prod.js gzip | 14.5 kB | 16.3 kB | ⚠️ +1.8 kB |
| app-route.ru...dev.js gzip | 21.7 kB | 23.5 kB | ⚠️ +1.82 kB |
| app-route.ru..prod.js gzip | 14.5 kB | 16.3 kB | ⚠️ +1.8 kB |
| pages-api-tu..prod.js gzip | 9.43 kB | 9.38 kB | N/A |
| pages-api.ru...dev.js gzip | 9.7 kB | 9.65 kB | N/A |
| pages-api.ru..prod.js gzip | 9.43 kB | 9.37 kB | N/A |
| pages-turbo...prod.js gzip | 22 kB | 21.9 kB | N/A |
| pages.runtim...dev.js gzip | 22.7 kB | 22.5 kB | N/A |
| pages.runtim..prod.js gzip | 22 kB | 21.9 kB | N/A |
| server.runti..prod.js gzip | 49.7 kB | 49.5 kB | N/A |
| Overall change | 103 kB | 114 kB | ⚠️ +10.9 kB |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>image-HASH.js</strong></summary>

```diff
@@ -1,7 +1,7 @@
 (self["webpackChunk_N_E"] = self["webpackChunk_N_E"] || []).push([
   [358],
   {
-    /***/ 1552: /***/ function (
+    /***/ 4070: /***/ function (
       __unused_webpack_module,
       __unused_webpack_exports,
       __webpack_require__
@@ -9,7 +9,7 @@
       (window.__NEXT_P = window.__NEXT_P || []).push([
         "/image",
         function () {
-          return __webpack_require__(2599);
+          return __webpack_require__(3697);
         },
       ]);
       if (false) {
@@ -18,7 +18,7 @@
       /***/
     },
 
-    /***/ 9272: /***/ function (module, exports, __webpack_require__) {
+    /***/ 2353: /***/ function (module, exports, __webpack_require__) {
       "use strict";
       /* __next_internal_client_entry_do_not_use__  cjs */
       Object.defineProperty(exports, "__esModule", {
@@ -40,15 +40,15 @@
         __webpack_require__(422)
       );
       const _head = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(8286)
+        __webpack_require__(2272)
       );
-      const _getimgprops = __webpack_require__(985);
-      const _imageconfig = __webpack_require__(8749);
-      const _imageconfigcontextsharedruntime = __webpack_require__(7829);
-      const _warnonce = __webpack_require__(3652);
-      const _routercontextsharedruntime = __webpack_require__(6120);
+      const _getimgprops = __webpack_require__(2530);
+      const _imageconfig = __webpack_require__(4899);
+      const _imageconfigcontextsharedruntime = __webpack_require__(9292);
+      const _warnonce = __webpack_require__(4213);
+      const _routercontextsharedruntime = __webpack_require__(7084);
       const _imageloader = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(5732)
+        __webpack_require__(7443)
       );
       // This is replaced by webpack define plugin
       const configEnv = {
@@ -373,7 +373,7 @@
       /***/
     },
 
-    /***/ 985: /***/ function (
+    /***/ 2530: /***/ function (
       __unused_webpack_module,
       exports,
       __webpack_require__
@@ -389,9 +389,9 @@
           return getImgProps;
         },
       });
-      const _warnonce = __webpack_require__(3652);
-      const _imageblursvg = __webpack_require__(1668);
-      const _imageconfig = __webpack_require__(8749);
+      const _warnonce = __webpack_require__(4213);
+      const _imageblursvg = __webpack_require__(3139);
+      const _imageconfig = __webpack_require__(4899);
       const VALID_LOADING_VALUES =
         /* unused pure expression or super */ null && [
           "lazy",
@@ -760,7 +760,7 @@
       /***/
     },
 
-    /***/ 1668: /***/ function (__unused_webpack_module, exports) {
+    /***/ 3139: /***/ function (__unused_webpack_module, exports) {
       "use strict";
       /**
        * A shared function, used on both client and server, to generate a SVG blur placeholder.
@@ -815,7 +815,7 @@
       /***/
     },
 
-    /***/ 9582: /***/ function (
+    /***/ 3067: /***/ function (
       __unused_webpack_module,
       exports,
       __webpack_require__
@@ -834,20 +834,24 @@
           });
       }
       _export(exports, {
-        getImageProps: function () {
-          return getImageProps;
+        unstable_getImgProps: function () {
+          return unstable_getImgProps;
         },
         default: function () {
           return _default;
         },
       });
       const _interop_require_default = __webpack_require__(1351);
-      const _getimgprops = __webpack_require__(985);
-      const _imagecomponent = __webpack_require__(9272);
+      const _getimgprops = __webpack_require__(2530);
+      const _warnonce = __webpack_require__(4213);
+      const _imagecomponent = __webpack_require__(2353);
       const _imageloader = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(5732)
+        __webpack_require__(7443)
       );
-      const getImageProps = (imgProps) => {
+      const unstable_getImgProps = (imgProps) => {
+        (0, _warnonce.warnOnce)(
+          "Warning: unstable_getImgProps() is experimental and may change or be removed at any time. Use at your own risk."
+        );
         const { props } = (0, _getimgprops.getImgProps)(imgProps, {
           defaultLoader: _imageloader.default,
           // This is replaced by webpack define plugin
@@ -860,9 +864,6 @@
             unoptimized: false,
           },
         });
-        // Normally we don't care about undefined props because we pass to JSX,
-        // but this exported function could be used by the end user for anything
-        // so we delete undefined props to clean it up a little.
         for (const [key, value] of Object.entries(props)) {
           if (value === undefined) {
             delete props[key];
@@ -877,7 +878,7 @@
       /***/
     },
 
-    /***/ 5732: /***/ function (__unused_webpack_module, exports) {
+    /***/ 7443: /***/ function (__unused_webpack_module, exports) {
       "use strict";
 
       Object.defineProperty(exports, "__esModule", {
@@ -912,7 +913,7 @@
       /***/
     },
 
-    /***/ 2599: /***/ function (
+    /***/ 3697: /***/ function (
       __unused_webpack_module,
       __webpack_exports__,
       __webpack_require__
@@ -933,8 +934,8 @@
 
       // EXTERNAL MODULE: ./node_modules/.pnpm/react@18.2.0/node_modules/react/jsx-runtime.js
       var jsx_runtime = __webpack_require__(1527);
-      // EXTERNAL MODULE: ./node_modules/.pnpm/file+..+main-repo+packages+next+next-packed.tgz_react-dom@18.2.0_react@18.2.0/node_modules/next/image.js
-      var next_image = __webpack_require__(1577);
+      // EXTERNAL MODULE: ./node_modules/.pnpm/file+..+diff-repo+packages+next+next-packed.tgz_react-dom@18.2.0_react@18.2.0/node_modules/next/image.js
+      var next_image = __webpack_require__(73);
       var image_default = /*#__PURE__*/ __webpack_require__.n(next_image); // CONCATENATED MODULE: ./pages/nextjs.png
       /* harmony default export */ var nextjs = {
         src: "/_next/static/media/nextjs.cae0b805.png",
@@ -964,12 +965,12 @@
       /***/
     },
 
-    /***/ 1577: /***/ function (
+    /***/ 73: /***/ function (
       module,
       __unused_webpack_exports,
       __webpack_require__
     ) {
-      module.exports = __webpack_require__(9582);
+      module.exports = __webpack_require__(3067);
 
       /***/
     },
@@ -980,7 +981,7 @@
       return __webpack_require__((__webpack_require__.s = moduleId));
     };
     /******/ __webpack_require__.O(0, [888, 774, 179], function () {
-      return __webpack_exec__(1552);
+      return __webpack_exec__(4070);
     });
     /******/ var __webpack_exports__ = __webpack_require__.O();
     /******/ _N_E = __webpack_exports__;
```
</details>
<details>
<summary>Diff for <strong>3f784ff6-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>68-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>polyfills-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>webpack-HASH.js</strong></summary>

```diff
@@ -219,7 +219,7 @@
     /******/ __webpack_require__.u = function (chunkId) {
       /******/ // return url for filenames based on template
       /******/ return (
-        "static/chunks/" + chunkId + "." + "6682446a0588d6cc" + ".js"
+        "static/chunks/" + chunkId + "." + "d0d10dabd8ac2b9e" + ".js"
       );
       /******/
     };
@@ -282,7 +282,6 @@
           /******/
         }
         /******/ script.setAttribute("data-webpack", dataWebpackPrefix + key);
-        /******/
         /******/ script.src = __webpack_require__.tu(url);
         /******/
       }
```
</details>
<details>
<summary>Diff for <strong>app-page-exp..ntime.dev.js</strong></summary>

```diff
failed to diff
```
</details>
<details>
<summary>Diff for <strong>app-page-exp..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.dev.js</strong></summary>

```diff
failed to diff
```
</details>
<details>
<summary>Diff for <strong>app-page.runtime.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-ex..ntime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-ex..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-tu..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-tu..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route.ru..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages-api-tu..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages-api.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages-api.ru..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages-turbo...time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages.runtime.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>server.runtime.prod.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 246a09a8f0c6878a19c82f442ad6d282392c32bc

**Review by styfle — COMMENTED** [review]
Thanks for the PR! 🎉 

Please add two new tests - one for hostname and one for pathname since both of those are changed.

The relevant test file is here: https://github.com/vercel/next.js/blob/543ca115716536b753549d950274da21ea979e64/test/unit/image-optimizer/match-remote-pattern.test.ts

Also the manifest file that we output from `next build` likely needs to be updated.

https://github.com/vercel/next.js/blob/be37d3eaabf67f418026af8b1e99b89558bc963b/packages/next/src/build/index.ts#L446-L452

Finally, its worth updating your branch since micromatch was recently changed to picomatch (maybe that already fixed it?)

- See https://github.com/vercel/next.js/pull/60699

**Comment by vordgi** [comment]
@styfle unfortunately, the bug is still there. However, now I understand that the solution above will only solve a specific bug of the tasks above, but there will still be situations like `pathname: "/base/**”`, where there will still be a path starting with a dot.

`picomatch` allows to pass the `{ dot: true }` setting, it helps, I think it's better to use it then.

Also, while testing now, I noticed that this error goes away, but the upstream image response error arrives for `https://bioage.typepad.com/.a/6a00d8341c4fbe53ef02c8d3a82122200d-600wi 403`

Looking at where it comes from…

**Comment by vordgi** [comment]
@styfle I checked where else there might be issues (_seems to be working in other places_) and saw comments on micromatch (_which is no longer used_). I think it's necessary to check how picomatch works and either update it or correct the comments (micromatch -> picomatch).

image

**Comment by vordgi** [comment]
As a result, I only enabled this option for paths.

I added tests for segments starting with a dot and also saw an unclosed case, when under `/*/` there may be several segments (`/base/*/file.png` for `/base/segment1/segment2/file.png`), also added checks

**Comment by vordgi** [comment]
LOL, I felt it!!!

image

**Comment by vordgi** [comment]
https://www.npmjs.com/package/picomatch#basic-globbing

> Matches any character zero or more times, including path separators. Note that `**` will only match path separators (/, and \\ with the windows option) when they are the only characters in a path segment. Thus, `foo**/bar` is equivalent to `foo*/bar`, and `foo/a**b/bar` is equivalent to `foo/a*b/bar`, and more than two consecutive stars in a glob path segment are regarded as a single star. Thus, `foo/***/bar` is equivalent to `foo/*/bar`

So picomatch doesn't work the way it's described in the next.js documentation:

https://nextjs.org/docs/app/api-reference/components/image#remotepatterns

> Wildcard patterns can be used for both pathname and hostname and have the following syntax:
> 
> \* match a single path segment or subdomain
> \*\* match any number of path segments at the end or subdomains at the beginning

@styfle I think it's better to comment out the tests for this and leave a todo for fix, as it doesn't directly relate to the current tasks. What do you think?

**Comment by vordgi** [comment]
_Turned off this test and left the todos associated with picomatch_

**Review by styfle — CHANGES_REQUESTED** [review]
The manifest file that we output from `next build` needs to be updated to match this same logic:

https://github.com/vercel/next.js/blob/be37d3eaabf67f418026af8b1e99b89558bc963b/packages/next/src/build/index.ts#L446-L452

File: test/unit/image-optimizer/match-remote-pattern.test.ts [hunk 1]
```diff
@@ -117,6 +117,8 @@ describe('matchRemotePattern', () => {
     expect(m(p, new URL('https://avatars.example.com'))).toBe(false)
     expect(m(p, new URL('https://avatars.sfo1.example.com'))).toBe(true)
     expect(m(p, new URL('https://avatars.iad1.example.com'))).toBe(true)
+    /** @todo fix the check so the test works */
+    // expect(m(p, new URL('https://avatars.iad1.iad2.example.com'))).toBe(false)
```
[the remark this answers is not in the document]
Comment (vordgi) [reply test/unit/image-optimizer/match-remote-pattern.test.ts:121]: Exactly. There is `.toBe(false)` and it failed 😬

[the remark this answers is not in the document]
Comment (vordgi) [reply test/unit/image-optimizer/match-remote-pattern.test.ts:121]:
```js
m({ hostname: 'avatars.*.example.com' }, new URL('https://avatars.iad1.iad2.example.com')) // true
```

[the remark this answers is not in the document]
Comment (styfle) [reply test/unit/image-optimizer/match-remote-pattern.test.ts:121]: Oh you're saying this is already broken, even without the changes from your PR. That is surprising 😳

**Review by styfle — APPROVED** [review]
Looks good, thanks for the PR! 🎉

I reverted the "todo" comments since we want to keep the commit history as simple as possible and only include the lines of code that need to change.

For the single asterisk bug, you can open a new issue to track it (and open a corresponding PR to fix it).

Thanks!

**Pull request merged**